ds0_updater.py

```diff
@@ -5,13 +5,36 @@
 import logger
 import conf
 
+# keep it super simple for now because we don't really implement --allowed-keys yet.
+def verify_latest_commit():
+    allowed_keys = ds0_utils.allowed_keys()
+
+    # ensure the latest commit is signed and verified
+    while True:
+        for key in allowed_keys:
+            logger.exe('gpg --import ' + key['file'], expectError=True)
+        output = logger.exe('git log --pretty="format:%G?" --show-signature HEAD^..HEAD')
+
+        if "Can't check signature" in output[0]:
+            logger.info('gpg keys cleared on startup. Trying again...')
+            continue
+
+        verified = False
+        for key in allowed_keys:
+            if key['rsa_check'] in output[0] and key['signature_check'] in output[0]:
+                verified = True
+        if not verified:
+            logger.error('Scripts using a non-signed commit. Please ensure commit is valid.')
+            logger.error('    If it was a missed signature, feel free to open a ticket at https://github.com/riptano/ComboAMI.')
+        break
+
 # Update the AMI codebase if it's its first boot
 if not conf.get_config("AMI", "CompletedFirstBoot"):
     # check if a specific commit was requested
     force_commit = ds0_utils.required_commit()
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```

Comment:
Why would the remote name bubble up to the user interface? IIUC, the remote name is what the repository is referred to as locally and isn't exposed when sharing URLs. So it's unclear to me when would `origin` (or any other name) be part of anything that one could pass to `--repository`.

I'm getting a bit confused so let's stick to examples, that is, closer to implementation. :) I assume our starting condition is a cloned official repository:
```
git clone https://github.com/riptano/ComboAMI
```
The default branch will be whatever the default is, right now `2.5`. So after the initial clone `git status` should print this:
```
$ git status
On branch 2.5
Your branch is up-to-date with 'origin/2.5'.

nothing to commit, working directory clean
$
```
At that point our bootstrapping script (`ds0_updater.py`) kicks in, parses the `--repository` argument (and sets defaults), and executes a bunch of git commands.

The `--repository` argument is a subset of what `npm` supports: a URL optionally followed by a `#` followed by a commit-ish which is "any tag, sha, or branch which can be supplied as an argument to `git checkout`." Examples:
```
--repository https://github.com/elad/ComboAMI
--repository https://github.com/elad/ComboAMI#my_tag
--repository https://github.com/elad/ComboAMI#my_branch
--repository https://github.com/elad/ComboAMI#e5e3d41fb5f12461509aa1b6079413b381930d81
```
The defaults can either be hard-coded or inferred from the initially cloned repository (easier maintenance I think). In our case they'd be:
```
repository = 'https://github.com/riptano/ComboAMI'
branch_or_tag = '2.5'
commit_id = None
```
The first git commands executed by `ds0_updater.py` reset the origin:
```
git remote remove origin
git remote add origin <repository>
git fetch
```
Now - if I understand correctly - we want to use `git reset --hard` to switch to the desired commit rather than `git pull` or `git checkout`. So first we translate our `branch_or_tag` to `commit_id` if the latter is still unset using `git show-ref`:
```
pattern = 'refs/remotes/origin/' + branch_or_tag + '|' + 'refs/tags/' + branch_or_tag
commit_id = git show-ref | grep -E <pattern> | cut -d' ' -f1
```
Finally we modify the workspace:
```
git reset --hard <commit_id>
```
That should work, but has one drawback: while we will have the correct revision, we won't *technically* be on the branch/tag the user specified. Personally I don't see much of a problem in that, even more so if it's documented.

Does the above match what you had in mind? It seems to work for me without exposing that `origin` is the remote name outside the internal scripts, but again I might be forgetting something. :)